docs/sdg/sdg-flow-yaml.md

```diff
@@ -0,0 +1,125 @@
+
+Compatibility rules
+
+1. If the major version of the YAML file is higher than the parser can handle, the parser should reject the file.
+2. If the minor version of the YAML file is higher than the highest version the parser is aware of, the parser should read the file but ignore any unrecognized content.
+3. If the file’s version is lower than the parser version, the parser should provide default values for any configuration introduced in later versions.
+
+Example parsing logic:
+
+```python
+def parse_custom_flow(content):
+    version = content['version']
+    major, minor = map(int, version.split('.'))
+
+    if major > PARSER_MAJOR:
+        raise IncompatibleVersionError("The custom flow file format is from a future major version.")
+    elif major <= PARSER_MAJOR and minor > PARSER_MINOR:
+        logger.warning("The custom flow file may have new features that will be ignored.")
+```
+
+### Flow Params
+
+The following parameters are supplied to a Flow constructor and used by the render() method:
+
+- client - an OpenAI completions API client for talking to the teacher model via the serving backend (i.e.  llama-cpp or vLLM)
+- model_family - e.g. mixtral or merlinite
+- model_id - a path name for the specific teacher model being used
+- num_iters - how many samples to generate
+- batched - whether the model serving backend supports batched mode
+
+For now, we assume that the `Block` classes know how to use these parameters and there is no need to use any sort of templating in the custom flows that could use these parameters.
+
+### Model Prompts
+
+Based on whether model_family is mixtral or merlinite, a different prompt is used with the teacher model
+
+```python
+_MODEL_PROMPT_MIXTRAL = "<s> [INST] {prompt} [/INST]"
+_MODEL_PROMPT_MERLINITE = "'<|system|>\nYou are an AI language model developed by IBM Research. You are a cautious assistant. You carefully follow instructions. You are helpful and harmless and you follow ethical guidelines and promote positive behavior.\n<|user|>\n{prompt}\n<|assistant|>\n'"
+```
+
+For now, we assume that the `Block` class will choose the appropriate model prompt based on the family and that there is no need to specify a custom prompt.
+
+### Prompt Config Files
+
+Every LLMBlock references a separate prompt config file, and presumably a custom pipeline will provide custom prompt configs too.
+
+These prompt config files are quite simple YAML files - they contain a single object with system, introduction, principles, examples, and generation keys. See e.g. src/instructlab/sdg/configs/skills/freeform_questions.yaml
+
+TODO: figure out how we want to handle these.
+
+### Model Serving
+
+Custom pipelines may have more unique model serving requirements. Instead of serving just one model, we may need to launch the model server with a model and an additional model with adapter.
+
+How should this be expressed in a flow configuration? Can it be inferred, or are explicit options needed?
+
```

Comment:
Here are some thoughts around integrating this into the CLI. There are some immediate steps and possible work we'd do in future iterations to improve the UX.

```suggestion
### CLI Integration

As of the current version of `ilab`, it supports `simple` and `full` as parameters to `--pipeline` to select one of the two types of built-in pipelines included in the library. Once we have support for loading custom pipelines, we need a way for these to be specified with the CLI. Here is a proposed CLI UX for this:

> `ilab data generate`

Use the default pipeline, `simple`.

> `ilab data generate --pipeline full`

Use the built-in `full` pipeline.

> `ilab data generate --pipeline full --pipeline-extend path/to/config.yaml`

Use the `full` pipeline, but extend it with the pipeline defined in the provided config file.

> `ilab data generate --pipeline none --pipeline-extend path/to/other_config.yaml`

Do not use a built-in pipeline. Use only the pipeline in the custom configuration.

### Future CLI Improvements

A possible improvement would be to have a well-defined place on the filesystem where custom pipeline configs can be automatically loaded and included as options to the `--pipeline` parameter.

For example, if the config format included new parameters, `name: full-extended` and `extends: full`, and the CLI discovered and loaded it automatically, we could support `--pipeline full-extended` without needing the additional `--pipeline-extend` option.

`/usr/share/instructlab/sdg/` is a proposed location for this as a place for a distribution of InstructLab to include pre-defined custom pipelines, at least for Linux. See the [Filesystem Hierarchy Standard](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch04s11.html) for more details on why this path is appropriate for this use-case.

It would also make sense to support a configuration directory for user's own custom pipeline configurations. Assuming there is a base config directory, these could go in a `sdg` subdirectory. There is a separate proposal that discusses a proposed configuration location: <https://github.com/instructlab/dev-docs/pull/104>. Note this is separate from the distribution-provided, read-only pipelines discussed above with a different location.
```